src/shared/creds-util.c

```diff
@@ -223,7 +224,11 @@ static int make_credential_host_secret(
         assert(dfd >= 0);
         assert(fn);
 
-        fd = openat(dfd, ".", O_CLOEXEC|O_WRONLY|O_TMPFILE, 0400);
+        /* For non-root users creating a temporary file using the openat(2) over "." will fail later, in the
+         * linkat(2) step at the end.  The reason is that linkat(2) requires the CAP_DAC_READ_SEARCH
+         * capability when it uses the AT_EMPTY_PATH flag. */
+        if (have_effective_cap(CAP_DAC_READ_SEARCH))
```

Comment:
have_effective_cap() possibly returns a negative error (in case the caps cannot be read for some reason). but you use it here as if it was strictly a boolean. please change to

```c
if (have_effective_cap(…) <= 0)
```

or so, so that it is clear to the reader how errors are handled here (they will trigger the fallback path now, which makes sense, I'd say.)

Reply:
should be done